morl_baselines/multi_policy/pcn/pcn.py

```diff
@@ -0,0 +1,441 @@
+        horizons = np.float32(horizons)
+        e_returns = []
+        for i in range(n):
+            transitions = self._run_episode(env, returns[i], np.float32(horizons[i] - 2), max_return)
+            # compute return
+            for i in reversed(range(len(transitions) - 1)):
+                transitions[i].reward += self.gamma * transitions[i + 1].reward
+            e_returns.append(transitions[0].reward)
+
+        e_returns = np.array(e_returns)
+        distances = np.linalg.norm(np.array(returns) - e_returns, axis=-1)
+        return e_returns, np.array(returns), distances
+
+    def save(self, filename: str = "PCN_model", savedir: str = "weights"):
+        """Save PCN."""
+        if not os.path.isdir(savedir):
+            os.makedirs(savedir)
+        th.save(self.model, f"{savedir}/{filename}.pt")
+
+    def train(
+        self,
+        env: gym.Env,
+        num_er_episodes: int = 500,
+        total_time_steps: int = 1e7,
+        num_step_episodes: int = 10,
+        num_model_updates: int = 100,
+        max_return: float = 250.0,
+        max_buffer_size: int = 500,
+        ref_point: np.ndarray = np.array([0.0, 0.0]),
+    ):
+        """Train PCN."""
+        self.global_step = 0
+        total_episodes = num_er_episodes
+        n_checkpoints = 0
+
+        # fill buffer with random episodes
+        self.experience_replay = []
+        for _ in range(num_er_episodes):
+            transitions = []
+            obs, _ = env.reset()
+            done = False
+            while not done:
+                action = env.action_space.sample()
+                n_obs, reward, terminated, truncated, _ = env.step(action)
+                transitions.append(Transition(obs, action, np.float32(reward).copy(), n_obs, terminated))
+                done = terminated or truncated
+                obs = n_obs
+                self.global_step += 1
+            # add episode in-place
+            self._add_episode(transitions, max_size=max_buffer_size, step=self.global_step)
+
+        while self.global_step < total_time_steps:
+            loss = []
+            entropy = []
+            for _ in range(num_model_updates):
+                l, lp = self.update()
+                loss.append(l.detach().cpu().numpy())
+                lp = lp.detach().cpu().numpy()
+                ent = np.sum(-np.exp(lp) * lp)
+                entropy.append(ent)
+
+            desired_return, desired_horizon = self._choose_commands(num_er_episodes)
+
+            # get all leaves, contain biggest elements, experience_replay got heapified in choose_commands
+            leaves_r = np.array([e[2][0].reward for e in self.experience_replay[len(self.experience_replay) // 2 :]])
+            # leaves_h = np.array([len(e[2]) for e in self.experience_replay[len(self.experience_replay) // 2 :]])
+
+            hv = hypervolume(ref_point, leaves_r)
+            hv_est = hv
+            self.writer.add_scalar("train/hypervolume", hv_est, self.global_step)
```

Comment:
Does this change a lot compared to the eval/hypervolume thingy?

Reply:
Yes, because this get the best trajectories in the buffer instead of evaluating the model.